### PR TITLE
refactor: move_manager/drifting/jetpack/glide movement.

### PR DESCRIPTION
**Pull request by warriorstar-orion**
## What Does This PR Do

This PR ports /tg/'s move_manager API, a replacement for use of the native BYOND `walk` procs. It introduces movement subsystems, which can be used for processing different kinds of movements, adding one for most atom movement and one for items drifting in space.

Two new components are introduced:

- A "space drift" component is introduced which registers atoms with the move manager and handles their continuing movement.
- A jetpack component is introduced, which cleans up jetpack code and coordinates application of space drift via signals depending on if the jetpack is stabilized or not.

The port comprises major pieces from these PRs:

- https://github.com/tgstation/tgstation/pull/62567
- https://github.com/tgstation/tgstation/pull/66628
- https://github.com/tgstation/tgstation/pull/76133

move_manager was initially implemented as an SS, but /tg/ appears to be moving away from allowing SSes which don't fire, so it is now a global datum `GLOB.move_manager` (just for ease of porting, I don't necessarily agree with the move).

Many signals are added.

Something which may be contentious is the glide_size_multiplier, a global which scales glide time based on TD. This may affect the appearance of gliding objects unrelated to movement but will require more testing.

I've had to step through a lot of this code line by line to get it to work, so if you think something needs more documentation, let me know and I should have a good enough understanding to add it.

I would have liked this to be more atomic but I couldn't find a good way to separate the various implementation work without distrusting the result.

## Why It's Good For The Game

As the original PR states:

> The walk() procs are effectively sleep loops...the MC gets a lot less time then it expects, because it's being consumed by a separate scheduler.

This puts control of movement in userspace and allows for more flexibility. As can be seen, multiple movement subsystems can exist in order to allow for prioritizing or different firing times (in this implementation, we have spacedrift as a separate manager versus everything else).

It's also a pseudo-prerequisite to basic mobs, again attempting to keep the work of porting that as atomic as possible by adding the beneficial dependencies separately.

For other benefits of this approach, see the code that's cleaned up in:

- `/mob/living/carbon/human/Process_Spacemove` which was coupled with modsuit jetpack behavior;
- not having to set `inertia_dir` on every item moving around;
- SSspacedrift not having to be a unique implementation but just another SSmovement
- removal of `sleep`-dependent code as in `/obj/item/extinguisher/mini/nozzle/afterattack`, `/obj/item/reagent_containers/spray`, and `/obj/effect/decal/cleanable/blood/gibs/proc/streak`, among others

## How it Works

Cribbing from the original PR again:

> We've got a selection of move_loop datums. Each type holds information about what's going to be moved, and how.
> So you've got a move_loop type for moving in a direction, two different flavors for moving towards something, one for moving randomly.
> Then we have a movement_packet datum. This tracks what move_loops are attached to an object, and which move loop is currently running.
> There's a few rules around move loops. Primarily, you can only have one move loop active at once. This is determined by the loops precedence. Secondly, you're only allowed to have one loop per subsystem.

TL;DR, the move_manager gets passed movement instructions attached to atoms in the form of move_loop datums and manages the work of moving things.

<details>

<summary>This provides mostly drop-in replacements for the `walk()` procs</summary>

- Removing something from the loop `walk(0)` -> `GLOB.move_manager.stop_looping()`
- Move in a direction `walk()` -> `GLOB.move_manager.move()`
- Move towards a thing, taking turf density into account`walk_to()` -> `GLOB.move_manager.move_to()`
- Move in a thing's direction, ignoring turf density `walk_towards()` -> `GLOB.move_manager.home_onto()` and `GLOB.move_manager.move_towards_legacy()`, check the documentation to see which you like better
- Move away from something, taking turf density into account `walk_away()` -> `GLOB.move_manager.move_away()`
- Move to a random place nearby. NOT random walk `walk_rand()` -> `GLOB.move_manager.move_rand()` is random walk, `GLOB.move_manager.move_to_rand()` is walk to a random place

</details>

It also provides some signals for handling the continuation and stopping of movement, which could be helpful in the future.

## Testing

Testing requires checking the behavior of everything that's had its `walk` proc usage replaced or which has been moved to the /drift component.

- [ ] Verify behavior for simplemob movement.
- [ ] Verify throwing things and collision in general.
- [ ] Verify behavior for hostile mobs in terms of movement towards/away targets and movement speeds.
- [ ] Verify behavior for use of fire extinguishers to propel wheelchairs.
- [ ] Verify behavior for spread of gib decals.
- [ ] Verify behavior for extinguisher nozzles.
- [ ] Verify behavior for spray bottles.
- [ ] Verify behavior for infrared beams on assemblies.
- [ ] Verify space drift in general.
- [ ] Verify coughing up blood in gravity and zero-g.
- [ ] Verify blood splatters in gravity and zero-g.
- [ ] Verify chances in wheelchair speed based on rider organ injuries/amputations.
- [ ] Verify behavior of things being shot through space via disposals, just in case.
- [ ] Verify behavior for jetpacks and movement when jetpacks are disabled.
- [ ] Verify same behavior for jetpack modsuit modules.
- [ ] Verify behavior for throwing objects in space and moving in opposite direction.
- [ ] Verify behavior for pushing off turfs and items in zero-g.
- [ ] Verify behavior for thrown items bouncing off turfs in zero-g.
- [ ] Verify behavior for items scattering from thrown trays.

This will require TMing especially when it comes to handling lots of moving things at once (e.g. a bunch of stuff drifting in space while someone is trying to fight Bubblegum), though, to be clear, I don't anticipate issues in such scenarios.

<hr>

### Declaration

- [x] I confirm that I either do not require [pre-approval](https://github.com/ParadiseSS13/Paradise/blob/master/CODE_OF_CONDUCT.md#types-of-changes-that-need-approval) for this PR, or I have obtained such approval and have included a screenshot to demonstrate this below.
  
  
  <hr>

## Changelog

Despite the relatively invasive nature of this change, this should have no player-facing changes, except for to improve performance. If there's any change or discrepancy in movespeed, throw speed, pull speed, projectiles, etc., it is a regression.

NPFC

### CODE REVIEW

File: code/modules/mod/modules/modules_general.dm [hunk 1]
```diff
@@ -128,40 +128,51 @@
...
+	.["stabilizers"] = add_ui_configuration("Stabilizers", "bool", stabilize)
 
 /obj/item/mod/module/jetpack/configure_edit(key, value)
 	switch(key)
 		if("stabilizers")
-			set_stabilizers(value)
+			configure_jetpack(value)
 
 /obj/item/mod/module/jetpack/proc/allow_thrust()
-	if(!active)
```
Comment (Qwertytoforty) [inline code/modules/mod/modules/modules_general.dm:148]: Stupid question, why not check for active here? Wouldn't the jetpack work without the modsuit being powered, or is there some other check?

**Comment by dj-34** [comment]
I'm wondering if this somehow fixes the diagonal movement issues
https://github.com/ParadiseSS13/Paradise/issues/16131

File: code/modules/mod/modules/modules_general.dm (hunk 1 above)
Comment (warriorstar-orion) [reply code/modules/mod/modules/modules_general.dm:148]: `allow_thrust` (i.e. the `thrust_callback` on the jetpack and which is passed as `check_on_move` to the component (yes it passes through 3 different but meaningful names unfortunately) is only invoked in `/datum/component/jetpack/proc/activate`, which is only called as a result of the `COMSIG_MODULE_TRIGGERED` signal firing in `/obj/item/mod/module/proc/on_activation`. And just before that, `active` is checked for us:

https://github.com/ParadiseSS13/Paradise/blob/5ca10363c1fedcf9a9c8326cc9cf612bd9f5eabc/code/modules/mod/modules/_modules.dm#L106-L110

**Comment by lewcc** [comment]
This is huge, thanks for putting in the work on it

**Comment by lewcc** [comment]
Testing some things out locally:

- Throwing a mob has some weird effects where they kind of glide for the last little bit of their movement
- Conveyors might move really slowly? I think jaunt might also, though I couldn't really verify if that was just a me thing or not
- I don't think slipping on banana peels works.

File: code/modules/mob/mob_movement.dm [hunk 2]
```diff
@@ -300,82 +316,126 @@
...
-	if(backup)
-		if(istype(backup) && movement_dir && !backup.anchored)
-			var/opposite_dir = turn(movement_dir, 180)
-			if(backup.newtonian_move(opposite_dir)) //You're pushing off something movable, so it moves
-				to_chat(src, "<span class='notice'>You push off of [backup] to propel yourself.</span>")
-		return 1
-	return 0
+		return TRUE
+
+	// TODO: if(buckled) may belong here
```
Comment (lewcc) [inline code/modules/mob/mob_movement.dm:330]: todo here

**Review by warriorstar-orion — COMMENTED** [review]
Closing this for now because https://github.com/ParadiseSS13/Paradise/pull/26762 has to be tested and merged first for this to work as expected, and I don't want this blocking the TM queue in the wrong order.
